Backend/lib/yudhishthira/src/Lib/Yudhishthira/Event/KaalChakra.hs

```diff
@@ -80,59 +86,89 @@ kaalChakraEventInternal h eventId req = withLogTag ("EventId-" <> eventId.getId)
       <> "; max batches: "
       <> show req.maxBatches
   chakraQueries <- QChakraQueries.findAllByChakra req.chakra
+  bn <- nextChakraBatchNumber req.chakra
+  let batchesNumber = bn - 1
+  batchedUserData <- fetchUserDataBatch req eventId chakraQueries batchesNumber
+  let usersNumber = length batchedUserData.userIds
+  endTime <- getCurrentTime
+  let durationInSeconds = Time.nominalDiffTimeToSeconds $ diffUTCTime endTime startTime
+  logInfo $ "User data fetched successfully: batches number: " <> show batchesNumber <> "; users number: " <> show usersNumber
+  logDebug $ "took time: " <> show durationInSeconds
+  pure $ Yudhishthira.RunKaalChakraJobRes (Just eventId) Nothing Nothing batchedUserData.chakraBatchState
+
+updateUserTagsHandler ::
+  (BeamFlow m r, CH.HasClickhouseEnv CH.APP_SERVICE_CLICKHOUSE m) =>
+  Handle m ->
+  Yudhishthira.UpdateKaalBasedTagsJobReq ->
+  m Yudhishthira.RunKaalChakraJobRes
+updateUserTagsHandler h req = do
+  if req.updateUserTags
+    then updateUserTagsHandlerInternal skipUpdateUserTagsHandler req
+    else updateUserTagsHandlerInternal h req
+
+updateUserTagsHandlerInternal ::
+  (BeamFlow m r, CH.HasClickhouseEnv CH.APP_SERVICE_CLICKHOUSE m) =>
+  Handle m ->
+  Yudhishthira.UpdateKaalBasedTagsJobReq ->
+  m Yudhishthira.RunKaalChakraJobRes
+updateUserTagsHandlerInternal h req = do
+  let eventId = req.eventId
+  startTime <- getCurrentTime
+  chakraQueries <- QChakraQueries.findAllByChakra req.chakra
   tags <- QNammaTag.findAllByChakra req.chakra
+
   -- Skip LLM tags instead of throwing error
   filteredTags <- flip filterM tags $ \tag -> case tag.rule of
     Yudhishthira.RuleEngine _ -> pure True
     Yudhishthira.LLM _ -> do
       logError $ "LLM is not implemented: tag: " <> tag.name <> "; skipping."
       pure False
 
-  loopData <- loopM (LoopData 0 S.empty req.batchDelayInSec) $ fetchUserDataBatch req eventId chakraQueries
-  let batchesNumber = loopData.batchNumber - 1
-  let usersNumber = length loopData.userIds
-  logInfo $ "User data fetched successfully: batches number: " <> show batchesNumber <> "; users number: " <> show usersNumber
+  bn <- nextChakraBatchNumber req.chakra
+  let batchNumber = bn - 1
+  let limit = Just $ req.usersInBatch
+  let offset = Just $ batchNumber * req.usersInBatch
+  batchedUserData <- QUserDataE.findAllByEventIdWithLimitOffset eventId limit offset
+  -- getting this, but maybe limit offset will leave some part of the data of trailing guyz begind so using this just as a medium to get the userIds to do tagging for.
```

Comment:
With this approach we will fetch user data twice, once with limit and offset, and then for each user separately. And update tags for some users twice (if data appeared twice). But you are right, we can use this query `findAllByEventIdWithLimitOffset` only for getting `userIds`, so let it be for now. If it affect performance, we'll think how to remove duplicacy

Reply:
yeah, i was also thinking this. Wanted to change this but saw that it was doing some findAll and was worried that with limit offset i am getting only userId and if a user had another entry just after the limit part then it might not be included and json logic might give wrong output.

Reply:
wanted to clarify with you this part that if findAll was even requried to fetch each individual users data or findOne itself should do ?